### ISSUE TITLE
Strange behavior of `Sequence`

### ISSUE THREAD

**Issue by masyagin1998** [keras-team/keras#11727]
Hello!
I am using latest tensorflow with GPU support and latest Keras.
I am trying to write parallel data augmentation using `keras.utils.Sequence`, but have two problems:
1) `fit_generator` creates too many workers (I have parameter `workers=2`, but it generates 4 additional workers);
2) Using `Sequence` is extremely slow (with generator and a lot of augmentations I spend two minutes on epoch, with multiprocessing and `Sequence` - more than 10 minutes.

Here is my `Sequence` and `fit_generator`:

`class ParallelDataGenerator(Sequence):
    """Generate images for training/validation/testing (parallel version)."""

    def __init__(self, dname: str, start: int, stop: int, batch_size: int, augmentate: bool):
        self.batch_size = batch_size
        self.augmentate = augmentate
        self.idxs = [i for i in range(start, stop)]
        self.imgs_in = [cv2.imread(os.path.join(dname, 'in', str(i) + '_in.png'),
                                   cv2.IMREAD_GRAYSCALE).astype(np.float32) for i in range(start, stop)]
        self.imgs_gt = [cv2.imread(os.path.join(dname, 'gt', str(i) + '_in.png'),
                                   cv2.IMREAD_GRAYSCALE).astype(np.float32) for i in range(start, stop)]
        self.idxs = [i for i in range(0, stop - start)]

    def __len__(self):
        return int(np.ceil(float(len(self.idxs)) / float(self.batch_size)))

    def on_epoch_end(self):
        shuffle(self.idxs)

    def __getitem__(self, idx):
        print(idx, os.getpid())
        start = idx * self.batch_size
        stop = start + self.batch_size
        if stop >= len(self.idxs):
            stop = len(self.idxs)

        imgs_in = []
        imgs_gt = []
        for i in range(start, stop):
            img_in = np.copy(self.imgs_in[self.idxs[i]])
            img_gt = np.copy(self.imgs_gt[self.idxs[i]])
            if self.augmentate:
                img_in, img_gt = random_effect_img(img_in, img_gt)
            imgs_in.append(normalize_in(img_in))
            imgs_gt.append(normalize_gt(img_gt))
        imgs_in = np.array(imgs_in)
        imgs_in.shape = (imgs_in.shape[0], imgs_in.shape[1], imgs_in.shape[2], 1)
        imgs_gt = np.array(imgs_gt)
        imgs_gt.shape = imgs_in.shape
        return imgs_in, imgs_gt`

`    model.fit_generator(
        train_generator,
        steps_per_epoch=(train_stop - train_start) / args.batchsize,
        epochs=args.epochs,
        validation_data=validation_generator,
        validation_steps=(validation_stop - validation_start) / args.batchsize,
        callbacks=callbacks,
        use_multiprocessing=True,
        workers=2,
        max_queue_size=10
    )`

**Comment by gabrieldemarmiesse** [comment]
Hello and thats for the report. Can you tell us how you know that 4 workers are used instead of two and also could you compare the performance with `workers=0`? Thanks.

**Comment by masyagin1998** [comment]
I used `HTOP` utility to check how many processes are running.

When running with `worker=0, use_multiprocessing=False`, I spend 3 minutes on epoch and have only one process in `HTOP`.

**Comment by gabrieldemarmiesse** [comment]
It's possible that the processes seem duplicated because of hyperthreading. (they'll use workers*2 virtual cores). For the fact that it's slower with multiprocessing, this is because there is the overhead of pickling since you don't do many operations on the worker side. There is also the overhead of starting the processes. This is linked to CPython and it's nothing we can really do about. It's a limitation of the `multiprocessing` library itself.
